### ISSUE TITLE
Implement, export, and document, a utility function for converting a file from FASTA to 2bit

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#5]
@kakopo Opening an issue for this just so that we have a convenient place to discuss things related to this first function that you're going to add to the **BSgenomeForge** package. We didn't discuss much of the details during our video call earlier today like function name, what arguments, and other details... 

Possible names for the function: `fastaTo2bit`, `fastaToTwoBit`, etc...

The function will have at least 2 arguments:
- One to specify the path to the original file (FASTA). Possible names for this argument: `srcfile`, `origfile`, `fromfile`, `fastafile`, etc...
- One to specify the path to the 2bit file to write. Possible name for this argument: `destfile`, `tofile`, `twobitfile`, etc...

This should be enough to implement the _basic functionality_ of converting from FASTA to 2bit. By _basic functionality_ I mean that for now the function should not try to be as smart as the `fasta_to_sorted_2bit.R` scripts that are located in the **BSgenome** package. For example:

- https://github.com/Bioconductor/BSgenome/blob/master/inst/extdata/GentlemanLab/BSgenome.Mfascicularis.NCBI.6.0-tools/fasta_to_sorted_2bit.R
- https://github.com/Bioconductor/BSgenome/blob/master/inst/extdata/GentlemanLab/BSgenome.Amellifera.NCBI.AmelHAv3.1-tools/fasta_to_sorted_2bit.R

So just focus on going directly from FASTA to 2bit for now, that is, don't try to rename or reorder the sequences, or to perform all the sanity checks that those `fasta_to_sorted_2bit.R` scripts are doing. We'll do this later.

**Dealing with IUPAC ambiguity letters (to do later):**

Genomic sequences sometimes contain a few IUPAC ambiguity codes. This is very rare though. Try to create a small toy FASTA file by hand with some ambiguity letters in it, and see what happens when you call `fastaTo2bit()` on it. One way to deal with this problem is to replace all ambiguity letters with `N`'s. This can be done with the `replaceAmbiguities()` function from **Biostrings**.

**Proposed improvement to the function (to do later):**

See issue #7.

**Comment by kakopo** [comment]
Hi @hpages. I've done a bit of rudimentary work but my code doesn't seem to work, or, at the very least, it throws the same errors I have written, as shown.
```
> fastaTo2bit("toyfasta2")
Error in fastaTo2bit("toyfasta2") : 
  'origfile' must be the path to an existing file
```
This is the code I currently have;
```
fastaTo2bit <- function(origfile="", destfile="")
{
  
if (!file.exists(origfile))
    stop(wmsg("'origfile' must be the path to an existing file"))
if (!file.exists(destfile))
    stop(wmsg("'destfile' must be the path to an existing file"))
if (file.access(destfile, 2))
    stop(wmsg("you don't have write permission to 'destfile'"))

file_dir <- file.path(origfile)
export.2bit(file_dir, destfile)

}
```
which shows where the error is from, but I'm having a hard time seeing where I went wrong. I would definitely appreciate some pointers on how to improve the function.

**Comment by lwaldron** [comment]
It looks like there is no file `toyfasta2` in the directory where you are running the script - do `getwd()` and `dir()` show what you expect?

**Comment by kakopo** [comment]
Hi @lwaldron. 
Both `getwd()` and `dir()` return the expected values- I am able to see the FASTA file I intend to convert, within the directory I am working in. And yet the error informs me otherwise.

**Comment by hpages** [comment]
Hi @kakopo ,

Very mysterious. Are you 100% sure that your `toyfasta2` file is in your working directory?

Your function is not working properly anyways, and I'll comment more on that, but at least the first sanity check should pass if `toyfasta2` is really in the current dir.

I just created a toy FASTA file (`toyfasta2.fa`, note that FASTA files should have the `.fa` or `.fsa` extension). After starting R in the same directory where the file is located, I get:
```
> fastaTo2bit("toyfasta2.fa")
Error in fastaTo2bit("toyfasta2.fa") : 
  'destfile' must be the path to an existing file
```
So I still get an error (and this error is expected), but it's no longer about `origfile`.

Did you try to do:
```
file.exists("toyfasta2")
```
after starting R in the same directory where your `toyfasta2` file is located? What do you get?

It seems to me that you are trying to follow too closely what was done with the `downloadGenomicSequencesFrom*()` functions. There are some important differences:

1. The 2 arguments here, `origfile` and `destfile`, are paths to files, not to directories. Even though, in the directory case, it makes sense to set a default value for the argument (usually `"."`, which designs the current directory), we can't do this with `origfile` and `destfile`. What would the defaults be? I see that you are setting these arguments to `""` by default, but the problem is that `""` is not a valid file name. So the question is: are there sensible defaults file paths that we could use for the input and output files? I don't think there are :disappointed: So we just have to live with the fact that the `origfile` argument is mandatory (i.e. no default value). And we're going to do the same for `destfile` (even though maybe there are ways to infer the path to the output file from the path to the input file, but let's keep things simple for now).

2. About these sanity checks:
    ```
    if (!file.exists(destfile))
        stop(wmsg("'destfile' must be the path to an existing file"))
    if (file.access(destfile, 2))
        stop(wmsg("you don't have write permission to 'destfile'"))
    ```
    The function will _create_ the output file (the 2bit file), so we should not expect this file to exist already.

3. Not clear what you are trying to achieve with:
    ```
    file_dir <- file.path(origfile)
    ```
    Please take the time to familiarize yourself with the `file.path()` function by consulting its man page. Make sure that you understand what it does. Try to use it in your R session by calling it on a string containing a file path, like you are doing with `file.path(origfile)`. Once you understand what the function does, you should be able to reassess the need for this line of code.

4. You cannot do `export.2bit(file_dir, destfile)`, this won't work. The `export.2bit()` function is defined and documented in the **rtracklayer** package. Check its man page to learn what it does and how to use it.

There will be a lot of man page reading. This is normal, don't worry :wink: 

H.

**Comment by kakopo** [comment]
Hi @hpages. Thank you for the feedback, and for the explanations over video call to get more in depth with what the problem required! I am currently working with this piece of code below. 
```
fastaTo2bit <- function(origfile, destfile)
{
myfile = readDNAStringSet(file.choose())

if (!file.exists(origfile))
    stop(wmsg("'origfile' must be the path to an existing file"))

export.2bit(myfile, "sortedfasta.2bit")

}
```
Of course, destfile is not explicitly put to use and the fasta file chosen is saved in the current working directory as `sortedfasta.2bit`, which I am currently working to improve. I found `file.choose()` helpful with regards to selecting the file directly via GUI, but it does not seem to return the `sortedfasta.2bit` file via CMD. I would like to know if this is because of `file.choose()` specifically, or something else in the code. Thank you!

**Comment by hpages** [comment]
You're making good progress.

`file.choose()` is an interesting function but it's not needed here because we're assuming that the user has _already_ supplied the path to the file via the `origfile` argument, so why would the function ask the user to supply this information again when it already has it?

`fastaTo2bit()` is a simple function. In its simplest version, that is, without argument checking, only 2 lines of code are needed in the body of the function to do the job.

**Comment by kakopo** [comment]
Hi @hpages. As always, I am grateful for the advice you share, as with this comment which helped me realize where I had gone off track. I have just opened a PR with the modified code, and I am looking forward to your feedback. Thanks!

**Comment by hpages** [comment]
Hi @kakopo, thanks for the PR. See https://github.com/Bioconductor/BSgenomeForge/pull/6#issuecomment-1345901284 for some feedback.

**Comment by hpages** [comment]
@kakopo I just merged PR #6. Also added you as an author of the package. Make sure that you resync your fork with the Bioconductor/BSgenomeForge repo. Let's move to issue #6. Thanks!
